lib/nlp/text_normalization.py

```diff
@@ -0,0 +1,196 @@
+        string2 = string2.decode('utf-8')
+
+    if len(string1) > len(string2):
+        string1, string2 = string2, string1
+    distances = list(range(len(string1) + 1))
+    for index2, char2 in enumerate(string2):
+        new_distances = [index2 + 1]
+        for index1, char1 in enumerate(string1):
+            if char1 == char2:
+                new_distances.append(distances[index1])
+            else:
+                new_distances.append(min((distances[index1] + substitution_cost,
+                                          distances[index1 + 1] + insertion_cost,
+                                          new_distances[-1] + deletion_cost)))
+        distances = new_distances
+        if max_distance and min(new_distances) > max_distance:
+            return max_distance
+
+    return distances[-1]
+
+
+def fit_text_to_format(input_text, regex_pattern, insert_edits=None):
+    """
+    Used to modify text to match the given regex pattern.
+    Args:
+        input_text (str): processed string with numerals and character constants fixed
+        regex_pattern (str): pattern to match
+        insert_edits (int): number of character deletes allowed for fuzzy matching
+
+    Returns:
+        input_text (str): modified text
+    """
+
+    if not insert_edits:
+        count = lambda l1, l2: sum([1 for x in l1 if x in l2])
+        insert_edits = count(input_text, PUNCTUATION_CHARACTERS) + 2
+
+    pattern = f'(?b)({regex_pattern}){{i<={insert_edits}}}'
+    pattern = re.compile(pattern, flags=_re_flags)
+    matched_format = pattern.search(input_text)
+
+    # Fuzzy matching acts in a non-greedy fashion, hence the following resolution of reverse iterations
+    range_matches = re.finditer(CAPTURE_RANGE_RE, regex_pattern)
+    for match in range_matches:
+        min_range = int(match["minimum"])
+        max_range = int(match["maximum"])
+        for i in range(max_range, min_range - 1, -1):
+            temp_pattern = regex_pattern.replace(match.group(), f'{{{i}}}')
+            pattern = f'(?b)({temp_pattern}){{i<={insert_edits}}}'
+            pattern = re.compile(pattern)
+            matched_format = pattern.search(input_text)
+            if matched_format:
+                regex_pattern = temp_pattern
+                break
+            if i == min_range:
+                return input_text
+
+    if matched_format:
+        if any(matched_format.fuzzy_counts):
+            fuzzy_edits = matched_format.fuzzy_changes[1]  # Insert edits are returned at position 1 in the tuple
+            for corrector, index in enumerate(sorted(fuzzy_edits, reverse=False)):
+                index -= corrector
+                input_text = _omit_character_by_index(input_text, index)
+    return input_text
+
+
+def _omit_character_by_index(text, index) -> str:
+    return text[:index] + text[index + 1:]
+
+
+def resolve_numerals(text) -> str:
+    """
+    Uses NumberDetector to resolve numeric occurrences in text for both English and Hindi.
```

Comment:
We have number detection algorithms for some other languages too besides en and hi at (ner_v2.detectors.numeral.number.en), is there a reason we are restricting it to only two?

Reply:
ASR support is only being built for English and Hindi in current scope (from Product team).

Reply:
Let me know what I am missing here. Hemant was recently working on IVR where he added support for multiple languages right, and I am assuming ASR is applied as a step in IVR.